### PR TITLE
rational: Implement approximation from floats and FromPrimitive for v…

### PR DESCRIPTION
**Pull request by sdroege**
…arious types

FromPrimitive is implemented for i8/16/32/64 and BigInt.

https://github.com/rust-num/num/issues/282

### CODE REVIEW

File: rational/src/lib.rs [hunk 1]
```diff
@@ -668,6 +668,164 @@ impl RatioErrorKind {
...
+    fn from_f64(n: f64) -> Option<Self> {
+        Ratio::from_float(n)
+    }
+}
+
+macro_rules! from_primitive_integer {
+    ($typ:ty) => {
+        impl FromPrimitive for Ratio<$typ> {
+            fn from_i64(n: i64) -> Option<Self> {
+                <$typ as NumCast>::from(n).map(|n| Ratio::new(n, One::one()))
```
Comment (cuviper) [inline rational/src/lib.rs]: Perhaps more directly as `$typ::from_i64(n).map(Ratio::from_integer)`?

File: rational/src/lib.rs [hunk 2]
```diff
@@ -668,6 +668,164 @@ impl RatioErrorKind {
...
+                Self::approximate_float(n)
+            }
+        }
+    }
+}
+
+from_primitive_integer!(i8);
+from_primitive_integer!(i16);
+from_primitive_integer!(i32);
+from_primitive_integer!(i64);
```
Comment (cuviper) [inline rational/src/lib.rs]: Might as well add `isize` at least, but why not any unsigned types?

File: rational/src/lib.rs [hunk 3]
```diff
@@ -668,6 +668,164 @@ impl RatioErrorKind {
...
+
+impl<T: Integer + Signed + Bounded + NumCast + Copy> Ratio<T> {
```
Comment (cuviper) [inline rational/src/lib.rs]: Can we reduce these bounds at all?  e.g. `Copy`, we usually only require `Clone`.

File: rational/src/lib.rs [hunk 4]
```diff
@@ -774,6 +932,15 @@ mod test {
...
     }
 
+    #[test]
+    fn test_approximate_float() {
+        assert_eq!(super::approximate_float(0.5f32, 1.0e-20, 30), Some(Ratio::new(1i64, 2)));
+        assert_eq!(super::approximate_float(0.5f64, 1.0e-20, 30), Some(Ratio::new(1i32, 2)));
+        assert_eq!(super::approximate_float(5f32, 1.0e-20, 30), Some(Ratio::new(5i64, 1)));
+        assert_eq!(super::approximate_float(5f64, 1.0e-20, 30), Some(Ratio::new(5i32, 1)));
+        assert_eq!(super::approximate_float(29.97f32, 1.0e-20, 30), Some(Ratio::new(2997i64, 100)));
+        assert_eq!(super::approximate_float(-29.97f32, 1.0e-20, 30), Some(Ratio::new(-2997i64, 100)));
```
Comment (cuviper) [inline rational/src/lib.rs]: Need some `None` tests in here.

File: rational/src/lib.rs (hunk 3 above)
Comment (sdroege) [inline rational/src/lib.rs]: Clone instead of Copy would be possible, yes. The others are all needed unless you can think of another way for the conversions between T and F and back

File: rational/src/lib.rs (hunk 4 above)
Comment (sdroege) [inline rational/src/lib.rs]: Good point! I'll update this later, thanks for the review

File: rational/src/lib.rs (hunk 2 above)
Comment (sdroege) [inline rational/src/lib.rs]: Adding u8 to u64, and usize/isize is not a problem. Will do so later.

However what would be missing then is a generic way of converting of unsigned integers due to the trait bound overlaps we discussed before.

File: rational/src/lib.rs (hunk 4 above)
Comment (sdroege) [inline rational/src/lib.rs]: Let me know if you want more tests. Might make sense to add something around quickcheck here at some point.

File: rational/src/lib.rs [hunk 5]
```diff
@@ -668,6 +668,176 @@ impl RatioErrorKind {
...
+// No Unsigned constraint because this also works on positive integers and is called
+// like that, see above
+fn approximate_float_unsigned<T, F>(val: F, max_error: F, max_iterations: usize) -> Option<Ratio<T>>
+    where T: Integer + Bounded + NumCast + Clone,
+          F: Float + NumCast
+{
+    // Continued fractions algorithm
+    // http://mathforum.org/dr.math/faq/faq.fractions.html#decfrac
+
+    assert!(val >= F::zero());
```
Comment (cuviper) [inline rational/src/lib.rs]: I think the unsigned `FromPrimitive` need to deal with negative values before they reach this assert!

File: rational/src/lib.rs (hunk 4 above)
Comment (cuviper) [inline rational/src/lib.rs]: How about some floating point boundaries?  NaN, infinity, and even the finite yet absurdly large and small.  And unsigned ratios need testing now too, including that behavior with negative floats.

I'll bet we could benefit from quickcheck throughout num!

File: rational/src/lib.rs (hunk 2 above)
Comment (cuviper) [inline rational/src/lib.rs]: Yeah, that limitation is OK for now.

File: rational/src/lib.rs (hunk 3 above)
Comment (cuviper) [inline rational/src/lib.rs]: Well, maybe using explicit `ToPrimitive` + `FromPrimitive` is slightly better.  Not everything that implements those will also implement `NumCast`, unfortunately.

It's a weird trait, e.g. that `NumCast` requires `Self: ToPrimitive`, and not `FromPrimitive`, but otherwise it seems like something that really ought to have a blanket `impl`.  (breaking changes...)

File: rational/src/lib.rs (hunk 5 above)
Comment (sdroege) [inline rational/src/lib.rs]: Not sure I understand what you mean here. The signed ones all go into approximate_float(), which calls approximate_float_unsigned() for the absolute value and then negates the result if the original value was negative. approximate_float_unsigned() asserts that the input value is >= 0, i.e. "unsigned".

Comment (cuviper) [inline rational/src/lib.rs]: For `<u64 as FromPrimitive>::from_f64(-1.0)`, I expect `None`.  Is that the case?

Comment (sdroege) [inline rational/src/lib.rs]: Ah I see now what you mean, thanks!

File: rational/src/lib.rs (hunk 4 above)
Comment (sdroege) [inline rational/src/lib.rs]: Done, quickcheck would be for another time

File: rational/src/lib.rs (hunk 5 above)
Comment (sdroege) [inline rational/src/lib.rs]: > For <u64 as FromPrimitive>::from_f64(-1.0), I expect None. Is that the case?

It is now, yes

**Comment by sdroege** [comment]
Are there any other changes you'd like to see?

**Comment by cuviper** [comment]
Looks really nice, thanks!

@homu r+
